tests/suites/test_suite_psa_crypto.data

```diff
@@ -5071,127 +5071,201 @@ derive_actions_without_setup:
 
```

Comment:
I have not reviewed the tests. Please make sure at least the following cases are tested for `PSA_ALG_TLS12_PSK_TO_MS`. (I have not attempted to be exhaustive.)

Good cases:
* [ ] Salt omitted.
* [ ] Salt passed with `input_bytes`, of reasonable length(s).
* [ ] Salt passed with `input_key`, of type `DERIVE`, of reasonable length(s).
* [ ] Salt passed with `key_derivation_key_agreement`.
* [ ] Empty salt passed with `input_bytes`.

Bad cases:
* [ ] Salt passed before seed.
* [ ] Salt passed after secret.
* [ ] Salt passed after label.
* [ ] Salt passed twice.
* [ ] Unreasonably large salt passed with `input_bytes`.
* [ ] Unreasonably large salt passed with `input_key`.
* [ ] Salt of type `RAW_DATA` passed with `input_key`.
* [ ] Pass a secret with `input_bytes` and a salt with `input_key`. `output_key` must fail with `NOT_PERMITTED`.

Reply:
I'm working on the new PR where we got additional input (other key) instead of salt for Mixed-PSK.
I stuck with implementation and tests for Mixed-PSK support `key_derivation_key_agreement`. 

According to the documentation:
```
For a DHE-PSKor ECDHE-PSK cipher suite the other secret should be the output of the `PSA_ALG_FFDH` or
`PSA_ALG_ECDH` key agreement performed with the peer. The recommended way to pass this input is to use a key 
derivation algorithm constructed as `PSA_ALG_KEY_AGREEMENT(ka_alg, PSA_ALG_TLS12_PSK_TO_MS(hash_alg))` and to 
call `psa_key_derivation_key_agreement()`.
```

I don't understand how this should work. 
`key_derivation_key_agreement` takes `private key` and `peer key` and produces shared secret that in case of mixed-psk should be the other secret, but how PSK should be provided ?

Reply:
I think the documentation is describing the the best way to do this, as it protects intermediate secrets, but that's a bit beyond our immediate goals for TLS. Right now, as an intermediate step, I think it's OK for TLS to keep using `psa_raw_key_agreement()` for the ECDH part, writing the result in the PMS buffer, and then later on get the input from that buffer when deriving the MS. Doing it "the right way" as recommended in the documentation may require non-trivial work such as changing the time when some computations are made, and is out of scope as far as #5663 is concerned, if that's what you were asking about.

Note: you can find an example of usage of `key_derivation_key_agreement` in `test_suite_psa_crypto.function`. (I've never used it myself, so I don't know the answer to you question, but hopefully the test code can help.)

Reply:
> how PSK should be provided

`psa_key_derivation_input_key` or `psa_key_derivation_input_bytes` with `PSA_KEY_DERIVATION_INPUT_SECRET`.

And the (EC)DHE shared secret is passed with `psa_key_derivation_key_agreement` or `psa_raw_key_agreement; psa_key_derivation_input_bytes` with `PSA_KEY_DERIVATION_OTHER_SECRET`.

Reply:
Thanks @mpg @gilles-peskine-arm.

So the full flow for mixed-psk with key agreement is:
- input seed,
- key agreement (input other key)
- input psk (secret)
- input label

Reply:
Got a question regarding last last bullet in bad cases:
> Pass a secret with input_bytes and a salt with input_key. output_key must fail with NOT_PERMITTED.
If the SALT input comes from a key but the SECRET input comes from input_bytes, output_key will not be permitted, since output_key only cares about where the SECRET input comes from (according to the specification). This is the right choice, so there is no need to change anything in the specification. Please make sure there's a unit test for it.

With my current implementation in this case (secret comes with from input bytes and other secret comes from input key)  key derivation is successful and test passes. I understand that in such scenario `psa_key_derivation_output_bytes()` should return `NOT_PERMITTED` ? Why this operation is not permitted?

Reply:
Schematically, a key derivation combines a secret input with public inputs and produces a potentially secret output. If you use `input_key` with a non-exportable key for the secret input and `output_key` for the secret output, the secret output cannot be reconstructed outside the crypto service. On the other hand, if you use `input_bytes` for the secret input, the output can be reconstructed outside the crypto service, by making the same calculation from the inputs. `output_key` is reserved for the case where the secret input comes from `input_key` to reduce the risk of doing a security review and thinking “my key comes from `output_key` so I know it can't be reconstructed outside the crypto service”, at least on systems where all keys are non-exportable.